test/unit/controller/project-group.controller.test.ts

```diff
@@ -0,0 +1,297 @@
+            expect(okSpy).toHaveBeenCalledWith(response, serviceResult);
+            expect(response.status).toHaveBeenCalledWith(200);
+            expect(response.send).toHaveBeenCalledWith(serviceResult);
+            expect(next).not.toHaveBeenCalled();
+        });
+
+        test("parses numeric pagination and type filters before delegating to the service", async () => {
+            const params = new ReferralCodeQueryParams();
+            const fromSpy = jest.spyOn(ReferralCodeQueryParams, "from").mockReturnValue(params);
+            const serviceResult: ReferralCodeListResult = {
+                total_items: 5,
+                per_page: 25,
+                current_page: 3,
+                total_pages: 1,
+                data: [],
+            };
+            const getSpy = jest.spyOn(referralCodeService, "getReferralCodes").mockResolvedValue(serviceResult);
+            const okSpy = jest.spyOn(httpResponses, "Ok");
+
+            const request = {
+                params: { projectGroupId: "proj-789" },
+                query: { page_no: "3", page_size: "25", type: "2" },
+            } as unknown as Request;
+            const response = createResponse();
+            const next = createNext();
+
+            await projectGroupController.getReferralCodes(request, response, next);
+
+            expect(fromSpy).toHaveBeenCalledWith(request.query);
+            expect(params.page_no).toBe(3);
+            expect(params.page_size).toBe(25);
+            expect(params.type).toBe(2);
+            expect(getSpy).toHaveBeenCalledWith("proj-789", params);
+            expect(okSpy).toHaveBeenCalledWith(response, serviceResult);
+            expect(response.status).toHaveBeenCalledWith(200);
+            expect(response.send).toHaveBeenCalledWith(serviceResult);
+            expect(next).not.toHaveBeenCalled();
+        });
+
+        test("routes service errors through the shared error handler", async () => {
+            const params = new ReferralCodeQueryParams();
+            jest.spyOn(ReferralCodeQueryParams, "from").mockReturnValue(params);
+            const error = new Error("boom");
+            jest.spyOn(referralCodeService, "getReferralCodes").mockRejectedValue(error);
+            const handleErrorSpy = jest.spyOn(Utility, "handleError").mockImplementation(() => undefined);
+
+            const request = {
+                params: { projectGroupId: "proj-001" },
+                query: {},
+            } as unknown as Request;
+            const response = createResponse();
+            const next = createNext();
+
+            await projectGroupController.getReferralCodes(request, response, next);
+
+            expect(handleErrorSpy).toHaveBeenCalledWith(response, next, error, "Error fetching referral codes.");
+        });
+    });
+
+    describe("createReferralCode", () => {
+        test("enriches the DTO with context before calling the service", async () => {
```

Comment:
add two more cases
- If code already exists, should return 402
- If code exists in the current records where isactive = false, should create the new one.